packages/backend-for-frontend/src/services/toolService.ts

```diff
@@ -0,0 +1,469 @@
+      keyId: jwt.header.kid,
+    },
+  });
+
+  const purposeId = unsafeBrandId<PurposeId>(jwt.payload.purposeId);
+
+  if (keyWithClient.client.kind === authorizationApi.ClientKind.enum.API) {
+    return {
+      errors: undefined,
+      data: {
+        clientKind: authorizationApi.ClientKind.enum.API,
+        kid: jwt.header.kid,
+        algorithm: "RS256",
+        publicKey: encodedPem,
+        clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
+        consumerId: unsafeBrandId<TenantId>(keyWithClient.client.consumerId),
+        purposeId,
+      },
+    };
+  }
+
+  const purpose = await purposeProcessClient.getPurpose({
+    params: { id: purposeId },
+    headers: ctx.headers,
+  });
+
+  const agreement = await retrieveAgreement(
+    agreementProcessClient,
+    purpose.consumerId,
+    purpose.eserviceId,
+    ctx
+  );
+
+  const descriptor = await retrieveDescriptor(
+    catalogProcessClient,
+    agreement.eserviceId,
+    agreement.descriptorId,
+    ctx
+  );
+
+  return {
+    errors: undefined,
+    data: {
+      clientKind: authorizationApi.ClientKind.enum.CONSUMER,
+      clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
+      kid: jwt.header.kid,
+      algorithm: "RS256",
+      publicKey: encodedPem,
+      purposeId,
+      consumerId: unsafeBrandId<TenantId>(keyWithClient.client.consumerId),
+      agreementId: unsafeBrandId<AgreementId>(agreement.id),
+      eServiceId: unsafeBrandId<EServiceId>(agreement.eserviceId),
+      agreementState: agreementStateToItemState(agreement.state),
+      purposeState: retrievePurposeItemState(purpose),
+      descriptorState: descriptorStateToItemState(descriptor.state),
+    },
+  };
+}
+
+async function retrieveAgreement(
+  agreementClient: AgreementProcessClient,
+  consumerId: string,
+  eserviceId: string,
+  ctx: WithLogger<BffAppContext>
+): Promise<agreementApi.Agreement> {
+  const agreements = await getAllAgreements(agreementClient, ctx.headers, {
+    consumersIds: [consumerId],
+    eservicesIds: [eserviceId],
+    states: [
+      agreementApi.AgreementState.Values.ACTIVE,
+      agreementApi.AgreementState.Values.SUSPENDED,
+    ],
+  });
+
+  assertOnlyOneAgreementForEserviceAndConsumerExists(
+    agreements,
+    eserviceId,
+    consumerId
+  );
+  return agreements[0];
+}
+
+async function retrieveDescriptor(
+  catalogClient: CatalogProcessClient,
+  eserviceId: string,
+  descriptorId: string,
+  ctx: WithLogger<BffAppContext>
+): Promise<catalogApi.EServiceDescriptor> {
+  const eservice = await catalogClient.getEServiceById({
+    params: { eServiceId: eserviceId },
+    headers: ctx.headers,
+  });
+
+  const descriptor = eservice.descriptors.find((d) => d.id === descriptorId);
+  if (!descriptor) {
+    throw eserviceDescriptorNotFound(eservice.id, descriptorId);
+  }
+
+  return descriptor;
+}
+
+function retrievePurposeItemState(purpose: purposeApi.Purpose): ItemState {
+  const activePurposeVersion = [...purpose.versions]
```

Comment:
we should take the last version with a state in `ACTIVE`, `SUSPENDED`, `ARCHIVED`.

We could centralize the state logic by converting all states with `purposeVersionStateToItemState`, taking the `ACTIVE` if exists, otherwise return `INACTIVE`

Reply:
Correct me if I'm wrong, but with the current sort + find and these states
- Version1: ARCHIVED
- Version2: ACTIVE
the function wouldn't return the Archived version?

Reply:
In this case it will just return the first one it finds. This differs from the retrieveAgreement as in that the active or suspended have priority. Should this be the logic also in this function?

Reply:
ok then yes, there can be multiple versions in state ARCHIVED, but only one in ACTIVE/SUSPENDED (which is the last or the second to last).
If the sort is ascending the version found would always be the ARCHIVED when present, and not the ACTIVE